src/domain/services/AICategoryService.ts

```diff
@@ -0,0 +1,64 @@
+import {AuthorizationContext} from "@/src/security/models/AuthorizationContext";
+import {AISecurityService} from "@/src/security/services/AISecurityService";
+import prismadb from "@/src/lib/prismadb";
+import {ForbiddenError} from "@/src/domain/errors/Errors";
+import {CategoryType} from "@prisma/client";
+import {BaseEntitySecurityService} from "@/src/security/services/BaseEntitySecurityService";
+import {SecuredResourceType} from "@/src/security/models/SecuredResourceType";
+import {SecuredAction} from "@/src/security/models/SecuredAction";
+import {SecuredResourceAccessLevel} from "@/src/security/models/SecuredResourceAccessLevel";
+
+export interface AICategoryTypeInterface {
+    aiId: string;
+    categoryType: CategoryType;
+}
+
+export class AICategoryService {
+    public async getAICategoryTypes(
+        authorizationContext: AuthorizationContext,
+        aiId: string): Promise<Array<AICategoryTypeInterface>> {
+        const hasReadPermission = AISecurityService.hasInstanceReadAccess(authorizationContext);
+        if (!hasReadPermission) {
+            throw new ForbiddenError("Forbidden");
+        }
+        const aiCategories : Array<AICategoryTypeInterface> = await prismadb.aICategoryType.findMany({where: {aiId: aiId}});
+        return aiCategories;
+    }
+
+    public async getAIsByCategoryType(
+        authorizationContext: AuthorizationContext,
+        categoryType: CategoryType) : Promise<Array<AICategoryTypeInterface>> {
+        const hasReadPermission = AISecurityService.hasInstanceReadAccess(authorizationContext);
+        if (!hasReadPermission) {
+            throw new ForbiddenError("Forbidden");
+        }
+        const aiCategories : Array<AICategoryTypeInterface> = await prismadb.aICategoryType.findMany({where: {categoryType: categoryType}});
+        return aiCategories;
+    }
+
+    public async performUpdateOrCreateAICategoryType(
+        authorizationContext: AuthorizationContext,
+        aiCategoryTypes : Array<AICategoryTypeInterface>) {
+        const hasWritePermission : boolean = BaseEntitySecurityService.hasPermission(
+            authorizationContext,
+            SecuredResourceType.AI,
+            SecuredAction.WRITE,
+            SecuredResourceAccessLevel.INSTANCE
+        );
+        if (!hasWritePermission) {
+            throw new ForbiddenError("Forbidden");
+        }
+        let areRecordsDeleted : boolean = false;
+        aiCategoryTypes.map(async (aiCategoryType) => {
+            const aiId : any = aiCategoryType.aiId;
+            const categoryType : CategoryType = aiCategoryType.categoryType;
+            if (!areRecordsDeleted) {
+                await prismadb.aICategoryType.deleteMany({where: {aiId: aiId}});
+                areRecordsDeleted = true;
+            }
+            await prismadb.aICategoryType.create({data : { aiId, categoryType }});
+        });
```

Comment:
I understand that here we are removing all categories for a specific AI and recreating them with the updated values. A couple of suggestions to this: 

1. Pass aiId as a parameter instead of part of the AICategoryTypeInterface array.
  - The method as is implies that you could create AI-Category mappings for different AIs and while all the mappings will get created, the deleteMany will only be executed for the first aiId. 
  - The aiId can be included as a path parameter in the API (POST api/v1/ai/[aiId]/categories)
  - The UI always calls this API for a single AI anyway, so making it as part of the path will make the behavior clearer and more explicit. 

2. Here's a suggested rewrite of the update code (assuming that aiId is passed as a method parameter)
```typescript
 await prismadb.$transaction(async (tx) => {
      await tx.aICategoryType.deleteMany({
        where: { aiId },
      });
      await tx.aICategoryType.createMany({
        data: aiCategoryTypes.map((aiCategoryType) => ({
          aiId: aiId,
          categoryType: aiCategoryType.categoryType,
        })),
        skipDuplicates: true,
      });
    });
```

The above executes both delete and create in a single transaction, to ensure we don't have any inconsistencies in the case of a partial failure. It also uses createMany instead of having multiple create statements. The number of categories is limited so multiple inserts wouldn't be a huge issue but still good to try to be as efficient as possible

Reply:
Added both in a single transaction